### ISSUE TITLE
codeQualityToolMain vs. codeQualityTool

### ISSUE THREAD

**Issue by Vampire** [gradle/gradle#8542]
As far as I remember, you usually leave out the `Main` for the `main` source set in task names, like `compileJava` vs. `compileTestJava` or `compileOtherSourceSetJava` and so on.
I just stumbled upon the the code quality plugins being different, being named `pmdMain`, `findbugsMain`, `jdependMain` and so on.
Maybe it would make sense to follow suit and also name the CQ task for the main source set just `codeQualityTool`?

It could be done backwards compatible e. g. by renaming the tasks to `codeQualityTool`, then add a simple no-op task called `codeQualityToolMain` that depends on `codeQualityTool` and emits a deprecation warning, then at some point it can be removed.
Alternatively to catch the guys that read no warnings and / or have it in some scripts as an intermediary step between deprecation warning and removal make the task fail the build so that a meaningful error message can be given and then later on remove it.

Just an idea of course, if you don't like it, just close it. :-)

**Comment by lacasseio** [comment]
That is a great idea, and would also like to see the naming align a bit more. The main idea that came out of native was task name should scale depending on the complexity of the project. Ignoring main from the name is definitely aligned with that. We should even be able to remove the language of the compile task (`compileJava` to `compile`) however that requires a lot more infra work which will happen further in the future. Your migration path seems great as well. Do you want to give it a try and provide a PR for that backward compatibility change?

**Comment by big-guy** [comment]
There's naming conflict between the tasks and the extension.  If you follow the `codeQualityTool` naming convention, then `checkstyle` is the task for the main source set, but Gradle's going to find the `CheckstyleExtension` first.  I think I would leave this alone.

**Comment by lacasseio** [comment]
Should the user configure the tasks or the extension? I understand the confusion and that would be a good reason to close. If the user should be configuring the task lazily now, then there shouldn't be a naming conflict. Just a thought, we can close if the confusion is too real for the users.

**Comment by big-guy** [comment]
Both, but preferably the user wouldn't care about the name.  e.g., instead of:

```
checkstyle {
   // configure extension
}
checkstyleMain {
   // configure the task for the 'main' source set
}
```
You would instead do...
```
checkstyle {
   // configure extension
   // new, hypothetical API à la BinaryCollection/TaskCollection
   tasks.withSourceSet(sourceSets.main).configureEach {
         // configure the task for the 'main' source set
   }
}
```

That way, you don't care about the name of the task. We could name it `checkstyleMainJavaDebugJdk13` if we had other variants of the same set of code.  The name is a crappy way to encode information that we have already and people may want to query against.

I could also see an argument that a task called `checkstyle` should actually run _all_ checkstyle checks in a project, not just the main one.

**Comment by lacasseio** [comment]
Right, good point. Looking closer at the code for code quality plugins, it seems they don't set any group for the task which would effectively mean they are implementation details.

@Vampire Could you explain your use case for invoking the task directly? Or even configuring the task directly? That could be a better way forward with understanding what users are trying to achieve and provide the right method on the DSL.

**Comment by Vampire** [comment]
Hm, good points I didn't really think about.
Regarding
> I could also see an argument that a task called `checkstyle` should actually run _all_ checkstyle checks in a project, not just the main one.

This maybe would make sense, but would also be inconsistent, `compile` does not call all compilation tasks. I don't say it shouldn't be done that way, just noting it will also be inconsistent, yet it might be good anyway.

> Could you explain your use case for invoking the task directly?

Well, why not?
For example I run the build, it fails the pmdWhatever task, I look at the report, fix the issues or change the config and then I don't want to run the full build, I just run the task to see whether it is green now and if so, run a full build again.

I also do things like `./gradlew -i --continue $(./gradlew --console=plain tasks --all | grep -i pmd | cut -d - -f 1)` to run all pmd tasks and only those when setting it up or changing its config or updating its version. For this a `pmd` task that runs all `pmd` tasks would actually be nice, I think I like that idea pretty much right now.

> Or even configuring the task directly?

Because it is necessary to do config that is not possible via the extension?
For example
```gradle
codenarc {
    toolVersion '1.3'
    // customized copy from http://codenarc.sourceforge.net/StarterRuleSet-AllRulesByCategory.groovy.txt
    // update with new rules when version is updated
    config resources.text.fromFile('config/codenarc/codenarc.groovy')
}
tasks.withType(CodeNarc) {
    compilationClasspath = sourceSets."${(it.name - ~/^codenarc/).uncapitalize()}".with {
        it.runtimeClasspath + it.compileClasspath + project.configurations."${it.compileOnlyConfigurationName}"
    }
    reports.xml.enabled true
}
```
to enable XML reports in addition to HTML reports and to fix the classpath that is empty for CodeNarc by default.
Or
```gradle
tasks.withType(SpotBugsTask) {
    classpath += sourceSets."${(it.name - ~/^spotbugs/).uncapitalize()}".with { it.compileClasspath + project.configurations."${it.compileOnlyConfigurationName}" }
    extraArgs '-projectName', "$project.name ($it.name)"
    extraArgs '-release', "$project.version"
    reports {
        xml.withMessages true
        html.stylesheet resources.text.fromFile('config/spotbugs/fancy-hist.xsl')
    }

    task "${it.name}HtmlReport" {
        def input = reports.xml.destination
        inputs.file reports.html.stylesheet.asFile() withPropertyName 'spotbugsStylesheet' withPathSensitivity NONE
        inputs.files fileTree(input) withPropertyName 'input' withPathSensitivity NONE skipWhenEmpty()
        def output = file(input.absolutePath.replaceFirst(/\.xml$/, '.html'))
        outputs.file output withPropertyName 'output'
        doLast {
            def factory = TransformerFactory.newInstance('net.sf.saxon.TransformerFactoryImpl', getClass().classLoader)
            def transformer = factory.newTransformer(new StreamSource(reports.html.stylesheet.asFile()));
            transformer.transform(new StreamSource(input), new StreamResult(output))
        }
    }
    it.finalizedBy "${it.name}HtmlReport"
}
```
to have XML and HTML report at the same time which SpotBugs and the SpotBugs Gradle plugin do not yet support, fix the classpath and provide extra arguments (the `-release` would be possible on the extension, but the `-projectName` is different for each task and the task configuration overwrites the extension configuration, so both need to be at the task level).

**Comment by Vampire** [comment]
btw. `in:codenarc-plugin` and `in:checkstyle-plugin` labels are missing
